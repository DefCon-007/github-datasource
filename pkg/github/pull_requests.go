package github

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/grafana/grafana-github-datasource/pkg/models"
	"github.com/grafana/grafana-plugin-sdk-go/data"
	"github.com/pkg/errors"
	"github.com/shurcooL/githubv4"
)

// QueryListPullRequests lists all pull requests in a repository
// {
//   search(query: "is:pr repo:grafana/grafana merged:2020-08-19..*", type: ISSUE, first: 100) {
//     nodes {
//       ... on PullRequest {
//         id
//         title
//       }
//   }
// }
type QueryListPullRequests struct {
	Search struct {
		Nodes []struct {
			PullRequest PullRequest `graphql:"... on PullRequest"`
		}
		PageInfo PageInfo
	} `graphql:"search(query: $query, type: ISSUE, first: 100, after: $cursor)"`
}

// PullRequest is a GitHub pull request
type PullRequest struct {
	Title  string
	State  string
	Author struct {
		User User `graphql:"... on User"`
	}
	Closed    bool
	IsDraft   bool
	Locked    bool
	Merged    bool
	ClosedAt  githubv4.DateTime
	CreatedAt githubv4.DateTime
	UpdatedAt githubv4.DateTime
	MergedAt  githubv4.DateTime
	Mergeable githubv4.MergeableState
	MergedBy  struct {
		User User `graphql:"... on User"`
	}
}

type PullRequests []PullRequest

func (p PullRequests) Frame() data.Frames {
	frame := data.NewFrame(
		"pull_requests",
		data.NewField("title", nil, []string{}),
		data.NewField("state", nil, []string{}),
		data.NewField("author_login", nil, []string{}),
		data.NewField("author_email", nil, []string{}),
		data.NewField("author_company", nil, []string{}),
		data.NewField("closed", nil, []bool{}),
		data.NewField("is_draft", nil, []bool{}),
		data.NewField("locked", nil, []bool{}),
		data.NewField("merged", nil, []bool{}),
		data.NewField("closed_at", nil, []time.Time{}),
		data.NewField("merged_at", nil, []time.Time{}),
		data.NewField("updated_at", nil, []time.Time{}),
		data.NewField("created_at", nil, []time.Time{}),
	)

	for _, v := range p {
		frame.AppendRow(
			v.Title,
			v.State,
			v.Author.User.Login,
			v.Author.User.Email,
			v.Author.User.Company,
			v.Closed,
			v.IsDraft,
			v.Locked,
			v.Merged,
			v.ClosedAt.Time,
			v.MergedAt.Time,
			v.UpdatedAt.Time,
			v.CreatedAt.Time,
		)
	}

	return data.Frames{frame}

}

// GetPullRequestsInRange uses the graphql search endpoint API to find pull requests in the given time range.
func GetPullRequestsInRange(ctx context.Context, client Client, opts models.ListPullRequestsInRangeOptions, from time.Time, to time.Time) (PullRequests, error) {
	search := strings.Join([]string{
		"is:pr",
		fmt.Sprintf("repo:%s/%s", opts.Owner, opts.Repository),
		fmt.Sprintf("%s:%s..%s", opts.TimeField.String(), from.Format(time.RFC3339), to.Format(time.RFC3339)),
	}, " ")

	var (
		variables = map[string]interface{}{
			"cursor": (*githubv4.String)(nil),
			"query":  githubv4.String(search),
		}

		pullRequests = []PullRequest{}
	)

	for {
		q := &QueryListPullRequests{}
		if err := client.Query(ctx, q, variables); err != nil {
			return nil, errors.WithStack(err)
		}
		prs := make([]PullRequest, len(q.Search.Nodes))

		for i, v := range q.Search.Nodes {
			prs[i] = v.PullRequest
		}

		pullRequests = append(pullRequests, prs...)

		if !q.Search.PageInfo.HasNextPage {
			break
		}
		variables["cursor"] = q.Search.PageInfo.EndCursor
	}

	return pullRequests, nil
}
