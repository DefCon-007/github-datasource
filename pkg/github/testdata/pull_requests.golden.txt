ðŸŒŸ This was machine generated.  Do not edit. ðŸŒŸ

Frame[0] 
Name: pull_requests
Dimensions: 14 Fields by 2 Rows
+----------------+----------------+--------------------+--------------------+----------------------+--------------+----------------+--------------+--------------+-----------------+--------------------+---------------------------------+---------------------------------+---------------------------------+
| Name: title    | Name: state    | Name: author_login | Name: author_email | Name: author_company | Name: closed | Name: is_draft | Name: locked | Name: merged | Name: mergeable | Name: closed_at    | Name: merged_at                 | Name: updated_at                | Name: created_at                |
| Labels:        | Labels:        | Labels:            | Labels:            | Labels:              | Labels:      | Labels:        | Labels:      | Labels:      | Labels:         | Labels:            | Labels:                         | Labels:                         | Labels:                         |
| Type: []string | Type: []string | Type: []string     | Type: []string     | Type: []string       | Type: []bool | Type: []bool   | Type: []bool | Type: []bool | Type: []string  | Type: []*time.Time | Type: []*time.Time              | Type: []time.Time               | Type: []time.Time               |
+----------------+----------------+--------------------+--------------------+----------------------+--------------+----------------+--------------+--------------+-----------------+--------------------+---------------------------------+---------------------------------+---------------------------------+
| PullRequest #1 | OPEN           | testUser           | user@example.com   | ACME corp            | false        | false          | false        | false        | MERGEABLE       | null               | null                            | 2020-08-25 16:21:56 +0000 +0000 | 2020-08-25 16:21:56 +0000 +0000 |
| PullRequest #2 | OPEN           | testUser2          | user2@example.com  | ACME corp            | true         | false          | false        | true         | MERGEABLE       | null               | 2020-08-25 18:21:56 +0000 +0000 | 2020-08-25 18:21:56 +0000 +0000 | 2020-08-25 16:21:56 +0000 +0000 |
+----------------+----------------+--------------------+--------------------+----------------------+--------------+----------------+--------------+--------------+-----------------+--------------------+---------------------------------+---------------------------------+---------------------------------+


====== TEST DATA RESPONSE (arrow base64) ======
FRAME=QVJST1cxAAD/////QAYAABAAAAAAAAoADgAMAAsABAAKAAAAFAAAAAAAAAEDAAoADAAAAAgABAAKAAAACAAAAFwAAAACAAAAKAAAAAQAAABI+v//CAAAAAwAAAAAAAAAAAAAAAUAAAByZWZJZAAAAGj6//8IAAAAGAAAAA0AAABwdWxsX3JlcXVlc3RzAAAABAAAAG5hbWUAAAAADgAAAEgFAADcBAAAcAQAAAQEAACYAwAAPAMAANgCAAB8AgAAIAIAALwBAABMAQAA1AAAAGwAAAAEAAAAAvv//xQAAABAAAAAQAAAAAAAAApAAAAAAQAAAAQAAADw+v//CAAAABQAAAAKAAAAY3JlYXRlZF9hdAAABAAAAG5hbWUAAAAAAAAAAML+//8AAAMACgAAAGNyZWF0ZWRfYXQAAGb7//8UAAAAQAAAAEAAAAAAAAAKQAAAAAEAAAAEAAAAVPv//wgAAAAUAAAACgAAAHVwZGF0ZWRfYXQAAAQAAABuYW1lAAAAAAAAAAAm////AAADAAoAAAB1cGRhdGVkX2F0AACe////FAAAAEAAAABAAAAAAAAKAUAAAAABAAAABAAAALj7//8IAAAAFAAAAAkAAABtZXJnZWRfYXQAAAAEAAAAbmFtZQAAAAAAAAAAiv///wAAAwAJAAAAbWVyZ2VkX2F0ABIAGAAUABMAEgAMAAAACAAEABIAAAAUAAAAQAAAAEgAAAAAAAoBSAAAAAEAAAAEAAAALPz//wgAAAAUAAAACQAAAGNsb3NlZF9hdAAAAAQAAABuYW1lAAAAAAAAAAAAAAYACAAGAAYAAAAAAAMACQAAAGNsb3NlZF9hdAAAAKr8//8UAAAAQAAAAEAAAAAAAAAFPAAAAAEAAAAEAAAAmPz//wgAAAAUAAAACQAAAG1lcmdlYWJsZQAAAAQAAABuYW1lAAAAAAAAAACU/P//CQAAAG1lcmdlYWJsZQAAAAr9//8UAAAAPAAAADwAAAAAAAAGOAAAAAEAAAAEAAAA+Pz//wgAAAAQAAAABgAAAG1lcmdlZAAABAAAAG5hbWUAAAAAAAAAAPD8//8GAAAAbWVyZ2VkAABi/f//FAAAADwAAAA8AAAAAAAABjgAAAABAAAABAAAAFD9//8IAAAAEAAAAAYAAABsb2NrZWQAAAQAAABuYW1lAAAAAAAAAABI/f//BgAAAGxvY2tlZAAAuv3//xQAAABAAAAAQAAAAAAAAAY8AAAAAQAAAAQAAACo/f//CAAAABQAAAAIAAAAaXNfZHJhZnQAAAAABAAAAG5hbWUAAAAAAAAAAKT9//8IAAAAaXNfZHJhZnQAAAAAGv7//xQAAAA8AAAAPAAAAAAAAAY4AAAAAQAAAAQAAAAI/v//CAAAABAAAAAGAAAAY2xvc2VkAAAEAAAAbmFtZQAAAAAAAAAAAP7//wYAAABjbG9zZWQAAHL+//8UAAAARAAAAEQAAAAAAAAFQAAAAAEAAAAEAAAAYP7//wgAAAAYAAAADgAAAGF1dGhvcl9jb21wYW55AAAEAAAAbmFtZQAAAAAAAAAAYP7//w4AAABhdXRob3JfY29tcGFueQAA2v7//xQAAABEAAAARAAAAAAAAAVAAAAAAQAAAAQAAADI/v//CAAAABgAAAAMAAAAYXV0aG9yX2VtYWlsAAAAAAQAAABuYW1lAAAAAAAAAADI/v//DAAAAGF1dGhvcl9lbWFpbAAAAABC////FAAAAEQAAABEAAAAAAAABUAAAAABAAAABAAAADD///8IAAAAGAAAAAwAAABhdXRob3JfbG9naW4AAAAABAAAAG5hbWUAAAAAAAAAADD///8MAAAAYXV0aG9yX2xvZ2luAAAAAKr///8UAAAAPAAAADwAAAAAAAAFOAAAAAEAAAAEAAAAmP///wgAAAAQAAAABQAAAHN0YXRlAAAABAAAAG5hbWUAAAAAAAAAAJD///8FAAAAc3RhdGUAEgAYABQAAAATAAwAAAAIAAQAEgAAABQAAABEAAAASAAAAAAAAAVEAAAAAQAAAAwAAAAIAAwACAAEAAgAAAAIAAAAEAAAAAUAAAB0aXRsZQAAAAQAAABuYW1lAAAAAAAAAAAEAAQABAAAAAUAAAB0aXRsZQAAAAAAAAD/////WAMAABQAAAAAAAAADAAWABQAEwAMAAQADAAAAGgBAAAAAAAAFAAAAAAAAAMDAAoAGAAMAAgABAAKAAAAFAAAADgCAAACAAAAAAAAAAAAAAAiAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAQAAAAAAAAACAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAEAAAAAAAAABAAAAAAAAAAAgAAAAAAAAASAAAAAAAAAAAAAAAAAAAAEgAAAAAAAAAEAAAAAAAAABYAAAAAAAAABgAAAAAAAAAcAAAAAAAAAAAAAAAAAAAAHAAAAAAAAAAEAAAAAAAAACAAAAAAAAAACgAAAAAAAAAqAAAAAAAAAAAAAAAAAAAAKgAAAAAAAAAEAAAAAAAAAC4AAAAAAAAABgAAAAAAAAA0AAAAAAAAAAAAAAAAAAAANAAAAAAAAAACAAAAAAAAADYAAAAAAAAAAAAAAAAAAAA2AAAAAAAAAAIAAAAAAAAAOAAAAAAAAAAAAAAAAAAAADgAAAAAAAAAAgAAAAAAAAA6AAAAAAAAAAAAAAAAAAAAOgAAAAAAAAACAAAAAAAAADwAAAAAAAAAAAAAAAAAAAA8AAAAAAAAAAQAAAAAAAAAAABAAAAAAAAGAAAAAAAAAAYAQAAAAAAAAgAAAAAAAAAIAEAAAAAAAAQAAAAAAAAADABAAAAAAAACAAAAAAAAAA4AQAAAAAAABAAAAAAAAAASAEAAAAAAAAAAAAAAAAAAEgBAAAAAAAAEAAAAAAAAABYAQAAAAAAAAAAAAAAAAAAWAEAAAAAAAAQAAAAAAAAAAAAAAAOAAAAAgAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAgAAAAAAAAACAAAAAAAAAAEAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAADgAAABwAAAAAAAAAUHVsbFJlcXVlc3QgIzFQdWxsUmVxdWVzdCAjMgAAAAAAAAAABAAAAAgAAAAAAAAAT1BFTk9QRU4AAAAACAAAABEAAAAAAAAAdGVzdFVzZXJ0ZXN0VXNlcjIAAAAAAAAAAAAAABAAAAAhAAAAAAAAAHVzZXJAZXhhbXBsZS5jb211c2VyMkBleGFtcGxlLmNvbQAAAAAAAAAAAAAACQAAABIAAAAAAAAAQUNNRSBjb3JwQUNNRSBjb3JwAAAAAAAAAgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAACQAAABIAAAAAAAAATUVSR0VBQkxFTUVSR0VBQkxFAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAAAAAAAAAAAAACoXhTilS4WAGjtslWPLhYAqF4U4pUuFgBo7bJVjy4WAGjtslWPLhYQAAAADAAUABIADAAIAAQADAAAABAAAAAsAAAAOAAAAAAAAwABAAAAUAYAAAAAAABgAwAAAAAAAGgBAAAAAAAAAAAAAAAAAAAAAAoADAAAAAgABAAKAAAACAAAAFwAAAACAAAAKAAAAAQAAABI+v//CAAAAAwAAAAAAAAAAAAAAAUAAAByZWZJZAAAAGj6//8IAAAAGAAAAA0AAABwdWxsX3JlcXVlc3RzAAAABAAAAG5hbWUAAAAADgAAAEgFAADcBAAAcAQAAAQEAACYAwAAPAMAANgCAAB8AgAAIAIAALwBAABMAQAA1AAAAGwAAAAEAAAAAvv//xQAAABAAAAAQAAAAAAAAApAAAAAAQAAAAQAAADw+v//CAAAABQAAAAKAAAAY3JlYXRlZF9hdAAABAAAAG5hbWUAAAAAAAAAAML+//8AAAMACgAAAGNyZWF0ZWRfYXQAAGb7//8UAAAAQAAAAEAAAAAAAAAKQAAAAAEAAAAEAAAAVPv//wgAAAAUAAAACgAAAHVwZGF0ZWRfYXQAAAQAAABuYW1lAAAAAAAAAAAm////AAADAAoAAAB1cGRhdGVkX2F0AACe////FAAAAEAAAABAAAAAAAAKAUAAAAABAAAABAAAALj7//8IAAAAFAAAAAkAAABtZXJnZWRfYXQAAAAEAAAAbmFtZQAAAAAAAAAAiv///wAAAwAJAAAAbWVyZ2VkX2F0ABIAGAAUABMAEgAMAAAACAAEABIAAAAUAAAAQAAAAEgAAAAAAAoBSAAAAAEAAAAEAAAALPz//wgAAAAUAAAACQAAAGNsb3NlZF9hdAAAAAQAAABuYW1lAAAAAAAAAAAAAAYACAAGAAYAAAAAAAMACQAAAGNsb3NlZF9hdAAAAKr8//8UAAAAQAAAAEAAAAAAAAAFPAAAAAEAAAAEAAAAmPz//wgAAAAUAAAACQAAAG1lcmdlYWJsZQAAAAQAAABuYW1lAAAAAAAAAACU/P//CQAAAG1lcmdlYWJsZQAAAAr9//8UAAAAPAAAADwAAAAAAAAGOAAAAAEAAAAEAAAA+Pz//wgAAAAQAAAABgAAAG1lcmdlZAAABAAAAG5hbWUAAAAAAAAAAPD8//8GAAAAbWVyZ2VkAABi/f//FAAAADwAAAA8AAAAAAAABjgAAAABAAAABAAAAFD9//8IAAAAEAAAAAYAAABsb2NrZWQAAAQAAABuYW1lAAAAAAAAAABI/f//BgAAAGxvY2tlZAAAuv3//xQAAABAAAAAQAAAAAAAAAY8AAAAAQAAAAQAAACo/f//CAAAABQAAAAIAAAAaXNfZHJhZnQAAAAABAAAAG5hbWUAAAAAAAAAAKT9//8IAAAAaXNfZHJhZnQAAAAAGv7//xQAAAA8AAAAPAAAAAAAAAY4AAAAAQAAAAQAAAAI/v//CAAAABAAAAAGAAAAY2xvc2VkAAAEAAAAbmFtZQAAAAAAAAAAAP7//wYAAABjbG9zZWQAAHL+//8UAAAARAAAAEQAAAAAAAAFQAAAAAEAAAAEAAAAYP7//wgAAAAYAAAADgAAAGF1dGhvcl9jb21wYW55AAAEAAAAbmFtZQAAAAAAAAAAYP7//w4AAABhdXRob3JfY29tcGFueQAA2v7//xQAAABEAAAARAAAAAAAAAVAAAAAAQAAAAQAAADI/v//CAAAABgAAAAMAAAAYXV0aG9yX2VtYWlsAAAAAAQAAABuYW1lAAAAAAAAAADI/v//DAAAAGF1dGhvcl9lbWFpbAAAAABC////FAAAAEQAAABEAAAAAAAABUAAAAABAAAABAAAADD///8IAAAAGAAAAAwAAABhdXRob3JfbG9naW4AAAAABAAAAG5hbWUAAAAAAAAAADD///8MAAAAYXV0aG9yX2xvZ2luAAAAAKr///8UAAAAPAAAADwAAAAAAAAFOAAAAAEAAAAEAAAAmP///wgAAAAQAAAABQAAAHN0YXRlAAAABAAAAG5hbWUAAAAAAAAAAJD///8FAAAAc3RhdGUAEgAYABQAAAATAAwAAAAIAAQAEgAAABQAAABEAAAASAAAAAAAAAVEAAAAAQAAAAwAAAAIAAwACAAEAAgAAAAIAAAAEAAAAAUAAAB0aXRsZQAAAAQAAABuYW1lAAAAAAAAAAAEAAQABAAAAAUAAAB0aXRsZQAAAGgGAABBUlJPVzE=
